### PR TITLE
mesa-quilt without quilt

### PR DESCRIPTION
**Pull request by sunweaver**
This PR works around various RPM based distro that lack the quilt utility in their build environments. In Debain this is not a problem, as quilt is one of the standard packaging tools.

Obviously, Fedora has quilt, too. For RHEL and SUSE, the mesa-quilt script now has a half-dumb fallback mode in case quilt is not available in $PATH.

### CODE REVIEW

**Comment by sunweaver** [comment]
@Ionic: please review this PR as paid work.

@uli42: @mikedep333: please try test builds in your various build envs. Thanks!

**Comment by sunweaver** [comment]
@uli42: this is not working yet... As you know I use a tiling window manager (i3). I can easily escape the input lock if auto-grab is off, for example.

Also, I wonder, if we shouldn't rather introduce "flags" as opposed to the "autograb on|off" and "input lock on|off" strings in the title. E.g. "nxagent [GIX]" (grab, input lock, xinerama on) vs. "nxagent [giX]" (grab and input off, Xinerama on). That makes the agent not seem soooo English.

Furthermore, the autograb mode standalone seems to work. So I wonder, if we should separate those two features for now and get the autograb code in as is.

The input lock, I can e.g. escape from when I disable the autograb mode and then manipulate the window (e.g. minize or some fancy tiling command in i3). The window moves by keyboard event to "somewhere" else, the mouse is left outside nxagent and nxagent still shows input lock. Not true.

So for input lock we maybe need to capture mouse-in and mouse-out events (seen from nxagent's perspective). If the user manages to escape from the nxagent windows, then a new mouse-in event should trigger the input lock once more and capture the pointer.

I can take a closer look tomorrow, but maybe the above gives you a clue and things are fixed by tomorrow morning. Hehehe...

**Comment by uli42** [comment]
The branch name is wrong ;-)

**Comment by sunweaver** [comment]
On  Mi 28 Jun 2017 10:27:08 CEST, Ulrich Sibiller wrote:

> The branch name is wrong ;-)

Yeah, I noticed too late. Let's ignore that...

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: I figure you want to do this (i.e., exporting the variables and moving into the other directory) only after checking for `[ -z "$QUILT_PATCHES" ]`?

Comment (Ionic) [reply mesa-quilt]: Although, strictly speaking, the old implementation was buggy in the sense that it only moved into the directory of `$QUILT_PATCHES` was unset, but always unconditionally moved to the previous directory in its stack, so we shouldn't blindly put that back.

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: `under stand`?

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: `grep -E -v "^#.*" < "../Mesa.patches/series" ...`

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: `... ! grep -q "$patchfile" < "../Mesa.patches/.applied-patches"`

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: `echo "$patchfile" >> "../Mesa.patches/.applied-patches"`

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: `sed -i` is not portable. :(

```bash
sed -e "/$patchfile/d" "../Mesa.patches/.applied-patches" > "../Mesa.patches/.applied-patches.tmp"
mv "../Mesa.patches/.applied-patches.tmp" "../Mesa.patches/.applied-patches"
```

File: nx-libs.spec
Comment (Ionic) [inline nx-libs.spec]: `application`?

**Review by Ionic — CHANGES_REQUESTED** [review]

File: mesa-quilt
Comment (uli42) [reply mesa-quilt]: Why quote the path here? It's a static string

File: mesa-quilt
Comment (uli42) [inline mesa-quilt:20]: Can you explain why we need this? We have two cases: 
1. we are in the base directory. Then the path to mesa is clear
2. we are not in the base directory. Then we can use $(dirname $0) to get to defined place

I don't see why climbing up until / gets hit makes any sense...

File: mesa-quilt
Comment (uli42) [inline mesa-quilt]: I've never seen anyone using 1> ...

File: mesa-quilt
Comment (uli42) [inline mesa-quilt]: use mkdir -p without the if

File: mesa-quilt
Comment (uli42) [inline mesa-quilt]: +1

File: mesa-quilt
Comment (uli42) [inline mesa-quilt]: @Ionic should be sed without `-i` here

Comment (uli42) [reply mesa-quilt]: One more thing: I prefer compact shell scripts, this one has far too many empty lines for me. But YMMV

**Review by uli42 — CHANGES_REQUESTED** [review]

File: mesa-quilt
Comment (Ionic) [reply mesa-quilt]: Still, it's never a bad idea to quote as much as possible.

**Review by Ionic — CHANGES_REQUESTED** [review]

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > I figure you want to do this (i.e., exporting the variables and moving into the other directory) only after checking for [ -z "$QUILT_PATCHES" ]?

The comment is actually a left over, I moved it further up and dropped the "unset QUILT_PATCHES" part.

Comment (sunweaver) [reply mesa-quilt]: > Although, strictly speaking, the old implementation was buggy in the sense that it only moved into the directory of $QUILT_PATCHES was unset, but always unconditionally moved to the previous directory in its stack, so we shouldn't blindly put that back.

Indeed.

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: under_stood!

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > `grep -E -v "^#.*" < "../Mesa.patches/series" ...`

Yes. Nice.

Comment (sunweaver) [reply mesa-quilt]: > Why quote the path here? It's a static string

@uli42: Quoting is _always_ good. It's become a habit...

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > `! grep -q "$patchfile" < "../Mesa.patches/.applied-patches"`

... Yes...

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > `echo "$patchfile" >> "../Mesa.patches/.applied-patches"`

Indeed.

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > One more thing: I prefer compact shell scripts, this one has far too many empty lines for me. But YMMV

@uli42: /me prefers readable scripts to compact scripts. I consider bash more as a progamming language than an admin tool, actually.

File: nx-libs.spec
Comment (sunweaver) [reply nx-libs.spec]: > `application`

Yes.

Using `... falling back to stupidly (un)applying our patches...` now...

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt:20]: > I don't see why climbing up until / gets hit makes any sense...

@uli42: I have a copy of the mesa-quilt script in ~/bin/. So I can use it from anywhere in the nx-libs tree and don't need to give the full path to the script either. Very handy.

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > I've never seen anyone using 1> ...

Oh... common practise here. Suppress standard out, but see errors if they occur. Also...

```
<command> > /dev/null
```

is a bashism afaik.

In dash, you have to do:

```
<command> 1>/dev/null 2>/dev/null
```

... at least at the time that dash was introduced.

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > use mkdir -p without the if

yes. Of course!

**Comment by sunweaver** [comment]
@Ionic: @uli42: I have considered all of your suggestions. I push more commits to this PR branch, with atomic changes addressing each comment of yours individually.

Before merging, the history of this PR branch should be flattend... Don't merge right away, please.

File: mesa-quilt
Comment (Ionic) [reply mesa-quilt]: Oh, on the contrary, that's perfectly POSIX-compliant behavior:

>[2.7.2 Redirecting Output](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18)
>
>The two general formats for redirecting output are:
>
>[n]>word
>[n]>|word
>
>where the optional n represents the file descriptor number. If the number is omitted, the redirection shall refer to standard output (file descriptor 1).

If dash failed to implement that correctly (and yet claimed compliance), then it was broken.

File: nx-libs.spec
Comment (Ionic) [reply nx-libs.spec]: Still here.

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: Can fail if the file does not exist, use `rm -f` instead to ignore a nonexistent file.

**Review by Ionic — CHANGES_REQUESTED** [review]
Also still using `sed -i`, which is non-portable.

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > Can fail if the file does not exist, use rm -f instead to ignore a nonexistent file.

Ack. Done.

Comment (sunweaver) [reply mesa-quilt]: > Still here. (appliance -> application)

Done

File: mesa-quilt
Comment (sunweaver) [reply mesa-quilt]: > sed -i is not portable. :(

Got that now.

**Comment by sunweaver** [comment]
@Ionic: next iteration for #470, please.

File: mesa-quilt
Comment (Ionic) [inline mesa-quilt]: Okay, much better.

There's still a big caveat, though: sed interprets `${patchfile}` as a regexp and there's no way to turn that off. We should be mostly fine if the filenames are sane, but once they contain chars like `[`, `]` and others, this will fail badly. Even without "evil" characters, each full stop character in filenames will be interpreted as "any character", which might lead to problems.

We could work around that, but it's too complicated for that PR. Will clean that up some time later.

**Review by Ionic — APPROVED** [review]
All done, squash and rebase against 3.6.x, please.

File: nx-libs.spec
Comment (Ionic) [inline nx-libs.spec]: It's `fedora`, not `fedora_version`.

**Review by Ionic — CHANGES_REQUESTED** [review]

**Review by Ionic — APPROVED** [review]

**Pull request merged**
